core/src/cluster_slot_state_verifier.rs

```diff
@@ -345,92 +348,120 @@ pub enum ResultingStateChange {
     SendAncestorHashesReplayUpdate(AncestorHashesReplayUpdate),
 }
 
-fn check_duplicate_confirmed_hash_against_frozen_hash(
+/// Checks the duplicate confirmed hash we observed against our local bank status
+///
+/// 1) If we haven't replayed locally do nothing
+/// 2) If our local is dead, mark for dump & repair
+/// 3) If our local is replayed but mismatch hash, notify fork choice of duplicate and dump &
+///    repair
+/// 4) If our local is replayed and matches hash, notify fork choice that we have the correct
+///    version
+fn check_duplicate_confirmed_hash_against_bank_status(
     state_changes: &mut Vec<ResultingStateChange>,
     slot: Slot,
     duplicate_confirmed_hash: Hash,
-    bank_frozen_hash: Hash,
-    is_dead: bool,
+    bank_status: BankStatus,
 ) {
-    if duplicate_confirmed_hash != bank_frozen_hash {
-        if is_dead {
+    match bank_status {
+        BankStatus::Unprocessed => {}
+        BankStatus::Dead => {
             // If the cluster duplicate confirmed some version of this slot, then
             // there's another version of our dead slot
             warn!(
                 "Cluster duplicate confirmed slot {} with hash {}, but we marked slot dead",
                 slot, duplicate_confirmed_hash
             );
-        } else {
+            state_changes.push(ResultingStateChange::RepairDuplicateConfirmedVersion(
```

Comment:
The original code would also still have pushed a `state_changes.push(ResultingStateChange::MarkSlotDuplicate(bank_frozen_hash));` right, which is probably why the `dead_state_update_2` test was failing?

Reply:
yep exactly, as per my comment https://github.com/solana-labs/solana/pull/31699/files#r1197061926 - I think marking a dead slot as duplicate (`MarkSlotDuplicate(Hash::default())`) is a no-op so it is safe to be removed.